### PR TITLE
Issues/555

### PR DESCRIPTION
**Pull request by OlegGordiichuk**
Refactoring of the tokens view controller.

### CODE REVIEW

File: Trust/Tokens/Views/TokenViewCell.swift [hunk 1]
```diff
@@ -102,4 +102,11 @@ class TokenViewCell: UITableViewCell {
 
         backgroundColor = viewModel.backgroundColor
     }
+    func flashBackground() {
+        backgroundView = UIView()
+        backgroundView!.backgroundColor = Colors.lightBlue
+        UIView.animate(withDuration: 2.0, animations: {
+            self.backgroundView!.backgroundColor = UIColor.white
```
Comment (vikmeup) [inline Trust/Tokens/Views/TokenViewCell.swift]: can you avoid force unwrap?

File: Trust/Tokens/ViewControllers/TokensViewController.swift [hunk 2]
```diff
@@ -96,53 +89,75 @@ class TokensViewController: UIViewController {
...
+    /// Token observation.
+    private func tokensObservation() {
+        viewModel.setTokenObservation { [weak self] (changes: RealmCollectionChange) in
+            guard let strongSelf = self else { return }
+            let tableView = strongSelf.tableView
+            switch changes {
+            case .initial:
+                tableView.reloadData()
+            case .update(_, let deletions, let insertions, let modifications):
+                tableView.beginUpdates()
```
Comment (vikmeup) [inline Trust/Tokens/ViewControllers/TokensViewController.swift]: I would split this logic for separate method: updateCall at index path

File: Trust/Tokens/Coordinators/TokensCoordinator.swift [hunk 3]
```diff
@@ -70,9 +71,6 @@ class TokensCoordinator: Coordinator {
                 return ERC20Token(contract: address, name: token.name, symbol: token.symbol, decimals: token.decimals)
             }()
             return newTokenViewController(token: token)
-        case .nonFungibleTokens:
```
Comment (hewigovens) [inline Trust/Tokens/Coordinators/TokensCoordinator.swift]: remove support for nonFungibleTokens?

Comment (vikmeup) [reply Trust/Tokens/Coordinators/TokensCoordinator.swift]: I'm gonna add support in the next few days.

File: Trust/Tokens/Network/TokensNetwork.swift [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+class TokensNetwork: TokensNetworkProtocol {
+    /// provider of a `TokensNetwork` to make network requests.
+    var provider: MoyaProvider<TrustService>
+    /// config of a `TokensNetwork` to have current configuration of the app.
+    var config: Config
+    /// balanceService of a `TokensNetwork` to obteint tokens balance.
+    var balanceService: TokensBalanceService
+    /// account of a `TokensNetwork` curent user account reference.
```
Comment (vikmeup) [inline Trust/Tokens/Network/TokensNetwork.swift]: can we use `let` and remove comments?

File: Trust/Tokens/Network/TokensNetwork.swift [hunk 5]
```diff
@@ -0,0 +1,99 @@
+// Copyright SIX DAY LLC. All rights reserved.
+
+import UIKit
+import Moya
+import TrustKeystore
+
+protocol TrustNetworkProtocol {
```
Comment (vikmeup) [inline Trust/Tokens/Network/TokensNetwork.swift]: can you move into separate file

File: Trust/Tokens/Storage/TokensDataStore.swift [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        try! realm.commitWrite()
+    }
+    /// ETH token constrcutor.
+    ///
+    /// - Parameters:
+    ///   - config: parameters of the token.
+    /// - Returns: `TokenObject` ETH token.
+    static func etherToken(for config: Config) -> TokenObject {
+        return TokenObject(
+            contract: "0x0000000000000000000000000000000000000000",
```
Comment (vikmeup) [inline Trust/Tokens/Storage/TokensDataStore.swift]: can we have chain ID into ether token? so we can distinguish that server side

File: Trust/Tokens/ViewControllers/TokensViewController.swift [hunk 7]
```diff
@@ -42,20 +36,20 @@ class TokensViewController: UIViewController {
         )
         return footer
     }()
-    let account: Wallet
     let tableView: UITableView
     let refreshControl = UIRefreshControl()
     weak var delegate: TokensViewControllerDelegate?
-
+    /// ethTimer of a `TokensViewController` to shedule eth balance update.
+    var ethTimer: Timer?
```
Comment (vikmeup) [inline Trust/Tokens/ViewControllers/TokensViewController.swift]: can you name it as `etherFetchTimer` so you don't need to have comments?

File: Trust/Tokens/ViewControllers/TokensViewController.swift [hunk 8]
```diff
@@ -96,53 +94,96 @@ class TokensViewController: UIViewController {
         refreshControl.beginRefreshing()
         fetch()
     }
-
     func fetch() {
         self.startLoading()
-        self.dataStore.fetch()
+        self.viewModel.updateTickers()
```
Comment (vikmeup) [inline Trust/Tokens/ViewControllers/TokensViewController.swift]: can we combine this three methods into one?

File: Trust/Transactions/Coordinators/BalanceCoordinator.swift [hunk 9]
```diff
@@ -28,28 +31,41 @@ class BalanceCoordinator {
...
+        ethTokenObservation = token.observe {[weak self] change in
+            switch change {
+            case .change(let properties):
+                for property in properties {
+                    guard property.name == "value", let property = property.newValue as? String else {
+                        return
+                    }
+                    self?.updateBalance(for: token, with: BigInt(property))
+                }
+            case .error(let error):
```
Comment (vikmeup) [inline Trust/Transactions/Coordinators/BalanceCoordinator.swift]: do we need to observer errors in here? 
I would just do case `.error, .deleted`

File: Trust/Tokens/Network/TokensNetwork.swift (hunk 4 above)
Comment (OlegGordiichuk) [reply Trust/Tokens/Network/TokensNetwork.swift]: Done

File: Trust/Tokens/Network/TokensNetwork.swift (hunk 5 above)
Comment (OlegGordiichuk) [reply Trust/Tokens/Network/TokensNetwork.swift]: Done

File: Trust/Tokens/Storage/TokensDataStore.swift (hunk 6 above)
Comment (OlegGordiichuk) [reply Trust/Tokens/Storage/TokensDataStore.swift]: I will do this on this issue https://github.com/TrustWallet/trust-wallet-ios/issues/415.

File: Trust/Tokens/ViewControllers/TokensViewController.swift (hunk 7 above)
Comment (OlegGordiichuk) [reply Trust/Tokens/ViewControllers/TokensViewController.swift]: Done

File: Trust/Tokens/ViewControllers/TokensViewController.swift (hunk 8 above)
Comment (OlegGordiichuk) [reply Trust/Tokens/ViewControllers/TokensViewController.swift]: Done

File: Trust/Transactions/Coordinators/BalanceCoordinator.swift (hunk 9 above)
Comment (OlegGordiichuk) [reply Trust/Transactions/Coordinators/BalanceCoordinator.swift]: Done

**Pull request merged**